arrabiata/src/interpreter.rs

```diff
@@ -0,0 +1,962 @@
+    ) -> Self::Variable;
+
+    /// Double the elliptic curve point given by the affine coordinates
+    /// `(x1, y1)` and save the result in the registers `pos_x` and `pos_y`.
+    fn double_ec_point(
+        &mut self,
+        pos_x: Self::Position,
+        pos_y: Self::Position,
+        x1: Self::Variable,
+        y1: Self::Variable,
+    ) -> (Self::Variable, Self::Variable);
+
+    /// Load the affine coordinates of the elliptic curve point currently saved
+    /// in the temporary accumulators. Temporary accumulators could be seen as
+    /// a CPU cache, an intermediate storage between the RAM (random access
+    /// memory) and the CPU registers (memory cells that are constrained).
+    ///
+    /// For now, it can only be used to load affine coordinates of elliptic
+    /// curve points given in the short Weierstrass form.
+    ///
+    /// Temporary accumulators could also be seen as return values of a function.
+    ///
+    /// # Safety
+    ///
+    /// No constraints are enforced. It is not also enforced that the
+    /// accumulators have been cleaned between two different gadgets.
+    unsafe fn load_temporary_accumulators(
+        &mut self,
+        pos_x: Self::Position,
+        pos_y: Self::Position,
+        side: Side,
+    ) -> (Self::Variable, Self::Variable);
+
+    /// Save temporary accumulators into the environment
+    ///
+    /// # Safety
+    ///
+    /// It does not have any effect on the constraints.
+    unsafe fn save_temporary_accumulators(
+        &mut self,
+        _v1: Self::Variable,
+        _v2: Self::Variable,
+        _side: Side,
+    );
+}
+
+/// Run the application
+pub fn run_app<E: InterpreterEnv>(env: &mut E) {
+    let x1 = {
+        let pos = env.allocate();
+        env.fetch_input(pos)
+    };
+    let _x1_square = {
+        let res = env.allocate();
+        env.square(res, x1.clone())
+    };
+}
+
+/// Run an iteration of the IVC scheme
+///
+/// It consists of the following steps:
+/// 1. Compute the hash of the public input.
+/// 2. Compute the elliptic curve addition.
+/// 3. Run the polynomial-time function.
+/// 4. Compute the hash of the output.
+/// The environment is updated over time.
+/// When the environment is the one described in the [Witness
+/// environment](crate::witness::Env), the structure will be updated
+/// with the new accumulator, the new public input, etc. The public output will
+/// be in the structure also. The user can simply rerun the function for the
+/// next iteration.
+/// A row must be created to generate a challenge to combine the constraints
+/// later. The challenge will be also accumulated over time.
+///
+/// FIXME: homogeneize
+/// FIXME: compute error terms
+/// FIXME: the resulting constraints do not include the selectors. We decided to
+/// use this design to only keep in the constraints the columns that are not
+/// fixed by the relation.
+pub fn run_ivc<E: InterpreterEnv>(env: &mut E, instr: Instruction) {
```

Comment:
I am confused here, I thought the instructions were building blocks to craft the IVC circuit, but the signature of this function make me understand that you create an IVC of one of those instruction

Reply:
We can change the interface later. `run_ivc` should execute each instruction, in order, to build the IVC circuit.